### PR TITLE
roc-toolkit: init at master

### LINKED ISSUE #129567: Roc Toolkit
**Project description**
Roc is a toolkit for real-time audio streaming over the network.

Basically, Roc is a network transport, highly specialized for the real-time streaming use case. The user writes the stream to the one end and reads it from another end, and Roc deals with all the complexity of the task of delivering data in time and with no loss. Encoding, decoding, adjusting rates, restoring losses -- all these are performed transparently under the hood.

The project is conceived as a swiss army knife for real-time streaming. It is designed to support a variety of network protocols, encodings, FEC schemes, and related features. The user can build custom configurations dedicated for specific use cases and choose an appropriate compromise between the quality, robustness, bandwidth, and compatibility issues.[1]

**Metadata**

* homepage URL: https://awesomeopensource.com/project/roc-streaming/roc-toolkit
* source URL: https://github.com/roc-streaming/roc-toolkit
* license: MPL-2.0
* platforms: any?

[1]: https://roc-streaming.org/toolkit/docs/about_project/overview.html

### PR DESCRIPTION
**Pull request by iclanzan**
I have not had success compiling the latest tagged release which is v0.1.5 from
April 2020. Latest master, with last commit dated Jun 15, compiles
successfully however. I disabled OpenFEC for this initial version since
I couldn’t get it to compile properly.

###### Motivation for this change

Closes https://github.com/NixOS/nixpkgs/issues/129567

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -4623,6 +4623,12 @@
     githubId = 3661115;
     name = "Ingo Blechschmidt";
   };
+  iclanzan = {
```
Comment (figsoda) [inline maintainers/maintainer-list.nix:4626]: this should be in a separate commit: `maintainers: add iclanzan`

File: pkgs/applications/audio/roc-toolkit/default.nix [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+, python
+, ragel
+, gengetopt
+, libuv
+, libpulseaudio
+, sox
+}:
+
+stdenv.mkDerivation {
+  name = "roc-toolkit";
```
Comment (figsoda) [inline pkgs/applications/audio/roc-toolkit/default.nix:14]:
```suggestion
  pname = "roc-toolkit";
  version = "unstable-2021-06-15";
```

File: pkgs/applications/audio/roc-toolkit/default.nix [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    "--disable-tests"
+    "--disable-doc"
+  ];
+
+  meta = with lib; {
+    description = "Toolkit for real-time audio streaming over the network";
+    homepage = "https://roc-streaming.org/";
+    license = licenses.mpl20;
+    platforms = platforms.linux ++ platforms.darwin;
+    maintainers = [ maintainers.iclanzan ];
```
Comment (figsoda) [inline pkgs/applications/audio/roc-toolkit/default.nix:41]:
```suggestion
    maintainers = with maintainers; [ iclanzan ];
```

File: pkgs/applications/audio/roc-toolkit/default.nix [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    "--disable-openfec"
+    "--disable-tests"
+    "--disable-doc"
+  ];
+
+  meta = with lib; {
+    description = "Toolkit for real-time audio streaming over the network";
+    homepage = "https://roc-streaming.org/";
+    license = licenses.mpl20;
+    platforms = platforms.linux ++ platforms.darwin;
```
Comment (figsoda) [inline pkgs/applications/audio/roc-toolkit/default.nix:40]: `platforms.all` could be good too
```suggestion
    platforms = platforms.unix;
```

File: pkgs/applications/audio/roc-toolkit/default.nix [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    maintainers = [ maintainers.iclanzan ];
+  };
+}
+
```
Comment (figsoda) [inline pkgs/applications/audio/roc-toolkit/default.nix:44]:
```suggestion
```

**Review by figsoda — CHANGES_REQUESTED** [review]

File: pkgs/applications/audio/roc-toolkit/default.nix [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+  buildInputs = [ libuv libpulseaudio sox ];
+
+  src = fetchFromGitHub {
+    owner = "roc-streaming";
+    repo = "roc-toolkit";
+    rev = "c89687330bfce6f4dce79826f7a235b581f2b49d";
+    sha256 = "1j2gqfpsh0hrwrwvanv6b893b9pds93rv7bl7am2050j3jqmfjrw";
+  };
+
+  prefixKey = "--prefix=";
```
Comment (catern) [inline pkgs/applications/audio/roc-toolkit/default.nix:26]: This is the default, you shouldn't to set this explicitly

**Comment by wedens** [comment]
I've found a derivation with the dependencies (including openfec) and pulseaudio modules:
https://github.com/vkleen/machines/blob/c3ca8298c6807b81565c4ca12f5f631616272edb/pkgs/applications/audio/misc/roc-toolkit/default.nix

Perhaps you can use something from that and credit the author or maybe the author is willing to contribute themselves ;)

/cc @vkleen

File: pkgs/applications/audio/roc-toolkit/default.nix [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+, sox
+}:
+
+stdenv.mkDerivation {
+  name = "roc-toolkit";
+
+  nativeBuildInputs = [ scons python ragel gengetopt ];
+  buildInputs = [ libuv libpulseaudio sox ];
+
+  src = fetchFromGitHub {
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/roc-toolkit/default.nix:19]: src goes before any build inputs.

File: pkgs/applications/audio/roc-toolkit/default.nix [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    owner = "roc-streaming";
+    repo = "roc-toolkit";
+    rev = "c89687330bfce6f4dce79826f7a235b581f2b49d";
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/roc-toolkit/default.nix:22]: Why not use https://github.com/roc-streaming/roc-toolkit/releases/tag/v0.1.5 ?

Comment (figsoda) [reply pkgs/applications/audio/roc-toolkit/default.nix:22]: > I have not had success compiling the latest tagged release which is v0.1.5 from
April 2020. Latest master, with last commit dated Jun 15, compiles
successfully however.

File: pkgs/applications/audio/roc-toolkit/default.nix (hunk 6 above)
Comment (iclanzan) [reply pkgs/applications/audio/roc-toolkit/default.nix:26]: The default is missing the `--` and compilation fails.

File: maintainers/maintainer-list.nix (hunk 1 above)
Comment (iclanzan) [reply maintainers/maintainer-list.nix:4626]: Just separate commit, or separate PR too?

Comment (figsoda) [reply maintainers/maintainer-list.nix:4626]: just separate commit, right before `roc-toolkit: init at unstable-2021-06-15`

**Comment by figsoda** [comment]
Result of `nixpkgs-review pr 139054` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>roc-toolkit</li>
  </ul>
</details>

**Review by figsoda — APPROVED** [review]
lgtm

**Comment by figsoda** [comment]
ofborg choked on darwin build, so I'm doing a rebuild
@ofborg build roc-toolkit

**Comment by figsoda** [comment]
`--disable-libunwind` didn't work on darwin and darwin build failed with
```c
src/modules/roc_core/target_libunwind/roc_core/backtrace.cpp:10:10: fatal error: 'libunwind.h' file not found
#include <libunwind.h>
         ^~~~~~~~~~~~~
```

**Comment by iclanzan** [comment]
Ok, I have included `libunwind` and also managed to get `openfec` to work thanks to the code referenced in [this comment](https://github.com/NixOS/nixpkgs/pull/139054#issuecomment-925664139).

File: pkgs/development/libraries/openfec/default.nix [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+    cp -d ../bin/Release/eperftool \
+      ../bin/Release/simple_client \
+      ../bin/Release/simple_server \
+      ../bin/Release/test_code_params \
+      ../bin/Release/test_create_instance \
+      ../bin/Release/test_encoder_instance \
+      $out/bin
+    cp -d ../bin/Release/libopenfec.so \
+      ../bin/Release/libopenfec.so.1 \
+      ../bin/Release/libopenfec.so.1.4.2 \
```
Comment (figsoda) [inline pkgs/development/libraries/openfec/default.nix:31]: replace `.so` with `stdenv.hostPlatform.extensions.sharedLibrary` should be able to fix the darwin build issue

**Comment by wozeparrot** [comment]
Any updates on this?

**Comment by bgamari** [comment]
Unfortunately I ended up unwittingly duplicating this work in #153087.

**Comment by iclanzan** [comment]
I am glad to see your PR got merged! I am going to close this then.
